File: DesignParrerns/app/src/main/java/com/example/designpatternsdemo/creationaldesignpattern/singletondp/Singleton.java
package com.example.designpatternsdemo.creationaldesignpattern.singletondp;

public class Singleton {
    private static volatile Singleton instance;
    private Singleton() {
        // private constructor to prevent instantiation
    }

    public static Singleton getInstance() {
        if(instance == null) {
            synchronized ((Singleton.class)) {
                // check again as multiple thread can reach above step
                if(instance == null) {
                    instance = new Singleton();
                }
            }
        }
        return  instance;
    }
}

File: DesignParrerns/app/src/main/java/com/example/designpatternsdemo/creationaldesignpattern/singletondp/Singleton.kt
package com.example.designpatternsdemo.creationaldesignpattern.singletondp

class Singleton {
    var data: String = "Default"


    companion object {
        //Lazy initialization with thread-safety
        val instance: Singleton by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
            Singleton()
        }
    }
}


class Singleton2{
    var data: String = "Default"

    @Volatile
    private  var instance: Singleton2? = null

    fun getSingleton(): Singleton2 {
        return instance ?: synchronized(this) {
            instance ?: Singleton2().also { instance = it }
        }
    }

}

object SingletonObject {
    var data: String = "Default"
}